ksql-examples/src/main/java/io/confluent/ksql/datagen/DataGen.java

```diff
@@ -58,19 +60,52 @@ static void run(final String... args) throws IOException {
       return;
     }
 
-    final Generator generator = new Generator(arguments.schemaFile, new Random());
     final Properties props = getProperties(arguments);
     final DataGenProducer dataProducer = ProducerFactory
         .getProducer(arguments.keyFormat, arguments.valueFormat, props);
+    final Optional<TokenBucket> tokenBucket = arguments.msgRate != -1
+        ? Optional.of(new TokenBucket(arguments.msgRate, arguments.msgRate)) : Optional.empty();
 
-    dataProducer.populateTopic(
-        props,
-        generator,
-        arguments.topicName,
-        arguments.keyName,
-        arguments.iterations,
-        arguments.maxInterval
-    );
+    final List<Thread> threads = new LinkedList<>();
+    for (int i = 0; i < arguments.numThreads; i++) {
+      threads.add(startProducerThread(arguments, dataProducer, props, tokenBucket));
+    }
+
+    for (final Thread t : threads) {
+      try {
+        t.join();
```

Comment:
What's the purpose of this call? I'm having trouble making sense of the docs.

Reply:
It blocks waiting for the thread to exit. Otherwise, the program will just exit without producing the records that the user asked it to.

Reply:
The only reason producer threads would exit is if they're interrupted, right? So `DataGen` will always exit with exit code 1? (Not saying there's anything wrong with this, just clarifying my understanding.)

Reply:
no they could just finish producing the requested number of records.

Reply:
Ah, good call. That means this PR also changes the meaning of `iterations` from the total number of messages produced to the number of messages produced per thread.